src/token/api/src/canister/is20_transactions.rs

```diff
@@ -1,78 +1,306 @@
-    // if it is less than or greater than ONE MINUTE WINDOW, we reject the
-    // transaction.
-    let created_at_time = match created_at_time {
+    let auction_balance = updates.balance_of(auction_account());
+    let updated_auction_balance = (auction_balance + auction_fee).ok_or(TxError::AmountOverflow)?;
+    updates.set_balance(auction_account(), updated_auction_balance);
+
+    // At this point all the checks are done and no further errors are possible, so we modify the
+    // canister state only at this point.
+
+    balances.apply_change(&updates);
+
+    Ok(())
+}
+
+fn validate_and_get_tx_ts(
+    canister: &impl TokenCanisterAPI,
+    caller: Principal,
+    transfer_args: &TransferArgs,
+) -> Result<u64, TxError> {
+    let now = ic::time();
+    let from = Account::new(caller, transfer_args.from_subaccount);
+    let to = transfer_args.to;
+
+    let created_at_time = match transfer_args.created_at_time {
         Some(created_at_time) => {
-            if now.abs_diff(created_at_time) > TX_WINDOW {
-                return Err(TxError::GenericError {
-                    message: "Created time is too far in the past or future".to_string(),
+            if now.saturating_sub(created_at_time) > TX_WINDOW {
+                return Err(TxError::TooOld {
+                    allowed_window_nanos: TX_WINDOW,
                 });
             }
+
+            if created_at_time.saturating_sub(now) > PERMITTED_DRIFT {
+                return Err(TxError::CreatedInFuture { ledger_time: now });
+            }
+
+            for tx in canister.state().borrow().ledger.iter().rev() {
+                if now.saturating_sub(tx.timestamp) > TX_WINDOW + PERMITTED_DRIFT {
+                    break;
+                }
+
+                if tx.timestamp == created_at_time
+                    && tx.from == from
+                    && tx.to == to
+                    && tx.memo == transfer_args.memo
+                    && tx.amount == transfer_args.amount
+                    && tx.fee == transfer_args.fee.unwrap_or(tx.fee)
+                {
+                    return Err(TxError::Duplicate {
+                        duplicate_of: tx.index,
+                    });
+                }
+            }
+
             created_at_time
         }
 
         None => now,
     };
 
-    let (fee, fee_to) = stats.fee_info();
-    let fee_ratio = bidding_state.fee_ratio;
+    Ok(created_at_time)
+}
 
-    if amount <= fee {
-        return Err(TxError::AmountTooSmall);
+pub fn mint(
+    state: &mut CanisterState,
+    caller: Principal,
+    to: Account,
+    amount: Tokens128,
+) -> TxReceipt {
+    let total_supply = state.balances.total_supply();
+    if (total_supply + amount).is_none() {
+        // If we allow to mint more then Tokens128::MAX then simplie operation such as getting
+        // total supply or token stats will panic, So we add this check to prevent this.
+        return Err(TxError::AmountOverflow);
     }
```

Comment:
The comment is redundant

```suggestion
(total_supply + amount).ok_or(TxError::AmountOverflow)?;
```

Reply:
Let me disagree on this one. Because I had to think whether I should add this check and what are the implications of not having it...

Reply:
How can `Tokens128` be more than `Tokens128::MAX`?
I don't think I know which part you are disagreeing on, could you elaborate? (I'm all for disagreement)

Reply:
Tokens128 cannot, but the total amount is combined of many users' balances, so if we don't check it, total balance can go over Tokens128::MAX.